File: 860-design-circular-queue/design-circular-queue.py
if (False):
     MyCircularQueue = Queue

class Queue():
    def __init__(self,k):
        # You must use Python List
        # Must use all spaces
        self._a = [None]*k #CANNOT CHANGE THIS. k space is already allocated
        self._MAX = k
        ## YOU CAN HAVE YOUR PRIVATE DATA MEMBER HERE
        self._front = 0
        self._rear = -1
        self._size = 0
        
    def enQueue(self, T)->'bool':
        ## YOU CANNOT CALL append in this routine as U already have enough space
        if self.isFull():
            return False
        ## YOU CAN DO: self._a[pos] = T #pos is some position between 0 to self._MAX-1        
        self._rear = (self._rear + 1) % self._MAX
        self._a[self._rear] = T
        self._size += 1
        return True
    
    def isFull(self):
        return self._size == self._MAX
    
    def isEmpty(self):
        return self._size == 0
        
    def Front(self)->'T':
        ## YOU CANNOT CALL pop(0). NOTE: pop(0) is O(n). We want THETA(1)
        if self.isEmpty():
            return -1
        return self._a[self._front]
        
    def Rear(self) ->'T':
        if self.isEmpty():
            return -1
        return self._a[self._rear]
    
    def deQueue(self) ->'bool':
        if self.isEmpty():
            return False
        self._a[self._front] = None
        self._front = (self._front + 1) % self._MAX
        self._size -= 1
        return True
        
    #Print from FRONT TO REAR
    def __str__(self)->'string':
        if self.isEmpty():
            return "Empty Queue"
        res = []
        i = self._front
        for _ in range(self._size):
            res.append(str(self._a[i]))
            i = (i + 1) % self._MAX
        return " -> ".join(res)        
    
    def __len__(self)->'int':
        return self._size
class MyCircularQueue(Queue):
    def __init__(self, k: 'int'):
        super().__init__(k)
    
    def enQueue(self, value: int) -> bool:
        return super().enQueue(value)
    
    def deQueue(self) -> bool:
        return super().deQueue()
    
    def Front(self) -> int:
        return super().Front()
    
    def Rear(self) -> int:
        return super().Rear()
    
    def isEmpty(self) -> bool:
        return super().isEmpty()
    
    def isFull(self) -> bool:
        return super().isFull()

File: 860-design-circular-queue/design-circular-queue.cpp
class MyCircularQueue {

private: 
    struct ListNode{
        int val;
        ListNode* next;
        ListNode* prev;
        ListNode(int v, ListNode* n = nullptr ,ListNode* p = nullptr): val(v),next(n),prev(p){}
    };

    int space;
    ListNode* left;
    ListNode* right;

public:
    MyCircularQueue(int k) {
        space = k;
        left = new ListNode(0);
        right = new ListNode(0,nullptr,left);
        left->next = right;
    }
    
    bool enQueue(int value) {
        if(isFull()) return false;
        ListNode* cur = new ListNode(value,right,right->prev);
        right->prev->next = cur;
        right->prev = cur;
        space--;
        return true;
    }
    
    bool deQueue() {
        if(isEmpty()) return false;
        ListNode* tmp = left->next;
        left->next = left->next->next;
        left->next->prev = left;
        delete tmp;
        space++;
        return true;
    }
    
    int Front() {
        return isEmpty() ? -1: left->next->val;
    }
    
    int Rear() {
        return isEmpty() ? -1: right->prev->val;
    }
    
    bool isEmpty() {
        return left->next == right;
    }
    
    bool isFull() {
        return space == 0;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */